units/src/fee.rs

```diff
@@ -146,19 +148,20 @@ impl FeeRate {
     /// fee is a non-integer amount, the amount is rounded up, ensuring that the transaction fee is
     /// enough instead of falling short if rounded down.
     ///
-    /// Returns [`None`] if overflow occurred.
-    #[must_use]
-    pub const fn checked_mul_by_weight(self, weight: Weight) -> Option<Amount> {
-        // No `?` operator in const context.
+    /// Returns [`NumOpResult::Error`] if overflow occurred.
+    pub const fn checked_mul_by_weight(self, weight: Weight) -> NumOpResult<Amount> {
         match self.to_sat_per_kwu().checked_mul(weight.to_wu()) {
-            Some(mul_res) => match mul_res.checked_add(999) {
-                Some(add_res) => match Amount::from_sat(add_res / 1000) {
-                    Ok(fee) => Some(fee),
-                    Err(_) => None,
-                },
-                None => None,
+            Some(fee) => match Amount::from_sat(fee) {
+                Ok(fee_amt) => match fee_amt.to_sat().checked_add(999) {
+                    Some(add_op) => match Amount::from_sat(add_op / 1_000) {
+                        Ok(ret_amt) => NumOpResult::Valid(ret_amt),
+                        _ => NumOpResult::Error(E(MathOp::Mul))
+                    }
+                    _ => NumOpResult::Error(E(MathOp::Mul))
+                }
+                _ => NumOpResult::Error(E(MathOp::Mul))
             },
-            None => None,
+            _ => NumOpResult::Error(E(MathOp::Mul)),
         }
```

Comment:
Perhaps this is cleaner?

```rust
        if let Some(fee) =  self.to_sat_per_kwu().checked_mul(weight.to_wu()) {
            if let Some(round_up) = fee.checked_add(999) {
                if let Ok(ret) = Amount::from_sat(round_up / 1_000) {
                    return NumOpResult::Valid(ret);
                }
            }
        }
        NumOpResult::Error(E::while_doing(MathOp::Mul))
```

Reply:
Yeah, we should give up functional style here for the sake of readability.

Reply:
Done.  I had actually preferred `if let` here before.